gk/fib.c

```diff
@@ -1732,6 +1759,170 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+ *   |____________|
+ */
+static int
+protected_read_grantor_lb_entries(lua_State *l)
+{
+	lua_Integer i, tbl_size;
+	struct ipaddr *gt_addrs;
+	struct ipaddr *gw_addrs;
+
+	tbl_size = lua_objlen(l, 2);
+	gt_addrs = *(struct ipaddr **)lua_touserdata(l, 4);
+	gw_addrs = *(struct ipaddr **)lua_touserdata(l, 5);
+
+	/* Iterate over a table of tables. */
+	for (i = 1; i <= tbl_size; i++) {
+		const char *gt_ip, *gw_ip;
+		int ret;
+
+		/* Get the table at index i. */
+		lua_rawgeti(l, 2, i);
+
+		/*
+		 * Make sure that the item inside
+		 * the table is a table itself.
+		 */
+		if (!lua_istable(l, 6))
+			luaL_error(l, "Grantor entry %ld is not a table\n", i);
+
+		lua_getfield(l, 6, "gt_ip");
+		lua_getfield(l, 6, "gw_ip");
+
+		gt_ip = luaL_checkstring(l, 7);
+		gw_ip = luaL_checkstring(l, 8);
+
+		ret = convert_str_to_ip(gt_ip, &gt_addrs[i - 1]);
+		if (ret < 0) {
+			luaL_error(l,
+				"Cannot convert Grantor IP %s to bytes\n",
+				gt_ip);
+		}
+
+		ret = convert_str_to_ip(gw_ip, &gw_addrs[i - 1]);
+		if (ret < 0) {
+			luaL_error(
+				l,
+				"Cannot convert gateway IP %s to bytes\n",
+				gw_ip);
+		}
+
+		/* Pop this Grantor/gateway IP pair and strings off stack. */
+		lua_pop(l, 3);
+	}
+
+	return 0;
+}
+
+static void
+add_grantor_entry_lb_verify_params(lua_State *l, const char **prefix,
+	lua_Integer *tbl_size, struct gk_config **gk_conf)
+{
+	uint32_t ctypeid;
+	uint32_t correct_ctypeid_gk_config = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_GK_CONFIG_PTR);
+	void *cdata;
+	size_t len;
+
+	if (lua_gettop(l) != 3) {
+		luaL_error(l,
+			"%s expected three arguments, however it received %d arguments\n",
+		      __func__, lua_gettop(l));
+	}
+
+	/* First argument must be a prefix string. */
+	*prefix = lua_tolstring(l, 1, &len);
+	if (*prefix == NULL || len == 0)
+		luaL_error(l, "Could not read prefix for adding load balanced Grantor set\n");
+
+	/* Second argument must be a table. */
+	luaL_checktype(l, 2, LUA_TTABLE);
+	*tbl_size = lua_objlen(l, 2);
+	if (*tbl_size <= 0)
+		luaL_error(l, "Table must have a positive number of Grantor entries\n");
+
+	/* Third argument must be of type CTYPE_STRUCT_GK_CONFIG_PTR. */
+	cdata = luaL_checkcdata(l, 3,
+		&ctypeid, CTYPE_STRUCT_GK_CONFIG_PTR);
+	if (ctypeid != correct_ctypeid_gk_config) {
+		luaL_error(l, "%s expected '%s' as the third argument\n",
+			__func__, CTYPE_STRUCT_GK_CONFIG_PTR);
+	}
+	*gk_conf = *(struct gk_config **)cdata;
+}
+
+int
+l_add_grantor_entry_lb(lua_State *l)
+{
+	const char *prefix;
+	struct ip_prefix prefix_info;
+	lua_Integer tbl_size;
+	struct gk_config *gk_conf;
+	struct ipaddr *gt_addrs;
+	struct ipaddr *gw_addrs;
+	struct ipaddr **ud;
+	const char *err_msg = NULL;
+	int ret;
+
+	/* Verify presence and types of parameters and read them in. */
+	add_grantor_entry_lb_verify_params(l, &prefix,
+		&tbl_size, &gk_conf);
+
+	ud = lua_newuserdata(l, sizeof(gt_addrs));
+	gt_addrs = rte_malloc("grantor_ips",
+		tbl_size * sizeof(*gt_addrs), 0);
```

Comment:
Allocate `tbl_size * sizeof(*gt_addrs)` bytes with `lua_newuserdata()`, so we don't need to call `rte_malloc()`, neither we need `lua_pcall()` later because the memory will be automatically release by Lua once `l_add_grantor_entry_lb()` returns normally or due to an error.

This comment also applies to `gw_addrs` a couple of lines down.